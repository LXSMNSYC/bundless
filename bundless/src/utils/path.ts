import path from 'path'
import defaultPathImpl from 'path'
import slash from 'slash'
import { cleanUrl } from './utils'

export const dotdotEncoding = '__..__'

// maybe keep track of namespace query here?
export function importPathToFile(
    root: string,
    request: string,
    pathImpl = defaultPathImpl,
) {
    if (!request) {
        return ''
    }
    request = decodeURIComponent(request)
    request = cleanUrl(request)
    request = removeLeadingSlash(request)
    request = request.replace(/__\.\.__/g, '..')
    request = pathImpl.resolve(root, request)
    return request
}

export function fileToImportPath(
    root: string,
    filePath: string,
    pathImpl = defaultPathImpl,
) {
    filePath = pathImpl.resolve(root, filePath)
    const relative = pathImpl.relative(root, filePath)
    filePath = slash(relative)
    filePath = filePath.replace(/\.\.(\/|\\)/g, dotdotEncoding + '$1')
    filePath = '/' + filePath
    return filePath
}

export function osAgnosticPath(absPath: string | undefined, root: string) {
    if (!root) {
        throw new Error(
            `root argument is required, cannot make os agnostic path for ${absPath}`,
        )
    }
    if (!absPath) {
        return ''
    }
    if (!path.isAbsolute(absPath)) {
        absPath = path.resolve(root, absPath)
    }
    return slash(path.relative(root, absPath))
}


export function removeLeadingSlash(p: string) {
    return p.startsWith('/') ? p.slice(1) : p
}