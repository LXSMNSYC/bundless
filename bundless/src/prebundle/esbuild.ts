import * as esbuild from 'esbuild'
import { Metafile } from 'esbuild'
import fromEntries from 'fromentries'
import fs from 'fs-extra'
import path from 'path'
import slash from 'slash'
import tmpfile from 'tmpfile'
import { Config, Platform } from '../config'
import { osAgnosticPath } from '../utils'
import * as plugins from '../plugins'
import {
    defaultImportableAssets as defaultImportableAssets,
    defaultLoader,
    isRunningWithYarnPnp,
    JS_EXTENSIONS,
    MAIN_FIELDS,
} from '../constants'
import { logger } from '../logger'
import { DependencyStatsOutput } from './stats'
import {
    OptimizeAnalysisResult,
    runFunctionOnPaths,
    stripColon,
} from './support'
import { PluginsExecutor } from '../plugins-executor'

export const commonEsbuildOptions = (
    config: Config = {},
): esbuild.BuildOptions => {
    const omitHashes = process.env.BUNDLESS_CONSISTENT_HMR_GRAPH_HASH != null
    return {
        target: 'es2020',
        entryNames: !omitHashes ? '[dir]/[name]-[hash]' : '[dir]/[name]',
        chunkNames: 'chunks/[name]-[hash]',
        minify: false,
        minifyIdentifiers: false,
        minifySyntax: false,
        metafile: true,
        minifyWhitespace: false,
        mainFields: MAIN_FIELDS,
        sourcemap: false,
        bundle: true,
        platform: 'browser',
        format: 'esm',
        write: true,
        logLevel: 'error',
        loader: {
            '.js': 'jsx',
            '.cjs': 'js',
            // '.svg': 'dataurl', // TODO enable svg as data uri in development and in build
            ...defaultLoader,
            ...config.loader,
        },
        define: generateDefineObject({ config }),
    }
}

export function generateDefineObject({
    config = {} as Config,
    platform = 'browser' as Platform,
    isProd = false,
}) {
    if (platform === 'node') {
        return {
            'process.browser': 'false',
            ...config.define, // TODO mock browser stuff like fetch? this allows me to target other platform like cloudflare workers ...
        }
    }
    const noop = 'String'
    const nodeEnv =
        process.env.NODE_ENV || (isProd ? 'production' : 'development')
    return {
        'process.env.NODE_ENV': JSON.stringify(nodeEnv),
        // ...generateEnvReplacements(config.env || {}),
        'process.pid': '0',
        // global: 'window',
        __filename: '""',
        __dirname: '""',
        // TODO remove defines and use inject instead
        // TODO use the process inject instead of define
        // process: '{}',
        global: 'window',
        // 'process.env': '{}',
        'process.browser': 'true',
        'process.version': '""',
        // 'process.argv': '[]',
        // module: '{}',
        // Buffer: noop,
        // 'process.cwd': noop,
        // 'process.chdir': noop,
        clearImmediate: noop,
        setImmediate: noop,
        ...config.define,
    }
}

export const defaultResolvableExtensions = [
    ...JS_EXTENSIONS,
    ...defaultImportableAssets,
    '.json',
    '.css',
]

export async function bundleWithEsBuild({
    entryPoints,
    root,
    dest: destLoc,
    config,
    ...options
}) {
    const { alias = {}, externalPackages = [], minify = false } = options

    const tsconfigTempFile = tmpfile('.json')
    await fs.promises.writeFile(tsconfigTempFile, makeTsConfig({ alias }))

    // rimraf.sync(destLoc) // do not delete or on flight imports will return 404

    const initialOptions: esbuild.BuildOptions = {
        entryPoints,
        ...commonEsbuildOptions(config),
        splitting: true, // needed to dedupe modules
        external: externalPackages,
        minify: Boolean(minify),
        minifyIdentifiers: Boolean(minify),
        minifySyntax: Boolean(minify),
        minifyWhitespace: Boolean(minify),
        mainFields: MAIN_FIELDS,
        tsconfig: tsconfigTempFile,
        sourcemap: 'inline',
        bundle: true,
        write: true,
        outdir: destLoc,
        metafile: true,
    }

    const executor = new PluginsExecutor({
        initialOptions,
        ctx: {
            config: { root },
            isBuild: true,
            root,
        },
        plugins: [
            ...(config.plugins || []),
            plugins.NodeGlobalsPolyfillPlugin({
                buffer: true,
                process: true,
                define: initialOptions.define,
            }),
            plugins.NodeModulesPolyfillPlugin({
                namespace: 'node-modules-polyfills',
            }),
            plugins.CssPlugin(),
            plugins.NodeResolvePlugin({
                name: 'prebundle-node-resolve',
                mainFields: MAIN_FIELDS,
                extensions: [
                    ...defaultResolvableExtensions,
                    ...(Object.keys(config.loader || {}) || []),
                ],
                onNonResolved: (p, importer, e) => {
                    logger.debug(e.message + '\n' + e.stack)
                    logger.warn(
                        `Cannot resolve '${p}' from '${importer}' during traversal, using yarn pnp: ${isRunningWithYarnPnp}`,
                    )
                },
            }),
            plugins.UrlResolverPlugin(),
        ],
    })

    const buildResult = await esbuild.build({
        ...initialOptions,
        plugins: executor.esbuildPlugins(),
    })

    await fs.promises.unlink(tsconfigTempFile)

    let meta = buildResult.metafile!
    meta = runFunctionOnPaths(meta, (p) => {
        p = stripColon(p) // namespace:/path/to/file -> /path/to/file
        return p
    })
    const esbuildCwd = process.cwd()
    const bundleMap = metafileToBundleMap({
        meta,
        esbuildCwd,
        root,
    })

    const analysis = metafileToAnalysis({ meta, root, esbuildCwd })

    const stats = metafileToStats({ meta, destLoc })

    return { stats, bundleMap, analysis }
}

function makeTsConfig({ alias }) {
    const aliases = Object.keys(alias || {}).map((k) => {
        return {
            [k]: [alias[k]],
        }
    })
    const tsconfig = {
        compilerOptions: { baseUrl: '.', paths: Object.assign({}, ...aliases) },
    }

    return JSON.stringify(tsconfig)
}

export type BundleMap = Partial<Record<string, string>>

/**
 * Returns aon object that maps from entry (relative path from root) to output (relative path from root too)
 */
export function metafileToBundleMap(_options: {
    root: string
    esbuildCwd: string
    meta: Metafile
}): BundleMap {
    const { meta, root, esbuildCwd } = _options

    const maps: Array<[string, string]> = Object.keys(meta.outputs)
        .map((output): [string, string] | undefined => {
            // chunks cannot be entrypoints
            const entry = meta.outputs[output].entryPoint
            if (!entry) {
                return
            }
            return [
                osAgnosticPath(path.resolve(esbuildCwd, entry), root),
                osAgnosticPath(path.resolve(esbuildCwd, output), root),
            ]
        })
        .filter(Boolean) as any

    const bundleMap = fromEntries(maps)

    return bundleMap
}

function metafileToAnalysis(_options: {
    meta: Metafile
    root: string
    esbuildCwd: string
}): OptimizeAnalysisResult {
    const { meta, root, esbuildCwd } = _options
    const analysis: OptimizeAnalysisResult = {
        isCommonjs: fromEntries(
            Object.keys(meta.outputs)
                .map((output): [string, true] | undefined => {
                    if (path.basename(output).startsWith('chunk.')) {
                        return
                    }
                    const info = meta.outputs[output]
                    if (!info) {
                        throw new Error(`cannot find output info for ${output}`)
                    }
                    const isCommonjs =
                        info.exports?.length === 1 &&
                        info.exports?.[0] === 'default'
                    if (!isCommonjs) {
                        return
                    }
                    // what if imported path ahs not yet been converted by prebundler? then prebundler should lock server, it's impossible
                    return [
                        osAgnosticPath(path.resolve(esbuildCwd, output), root),
                        isCommonjs,
                    ]
                })
                .filter(Boolean) as any,
        ),
    }
    return analysis
}

export function metafileToStats(_options: {
    meta: Metafile
    destLoc: string
}): DependencyStatsOutput {
    const { meta, destLoc } = _options
    const stats = Object.keys(meta.outputs).map((output) => {
        const value = meta.outputs[output]
        // const inputs = meta.outputs[output].bytes;
        return {
            path: output,
            isCommon: ['chunk.'].some((x) =>
                path.basename(output).startsWith(x),
            ),
            bytes: value.bytes,
        }
    })

    function makeStatObject(value) {
        const relativePath = slash(path.relative(destLoc, value.path))
        return {
            [relativePath]: {
                size: value.bytes,
                // gzip: zlib.gzipSync(contents).byteLength,
                // brotli: zlib.brotliCompressSync ? zlib.brotliCompressSync(contents).byteLength : 0,
            },
        }
    }

    return {
        common: Object.assign(
            {},
            ...stats.filter((x) => x.isCommon).map(makeStatObject),
        ),
        direct: Object.assign(
            {},
            ...stats.filter((x) => !x.isCommon).map(makeStatObject),
        ),
    }
}
