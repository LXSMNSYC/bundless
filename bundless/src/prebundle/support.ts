import { Metadata } from 'esbuild'
import { EventEmitter, once } from 'events'
import { forOwn, isPlainObject } from 'lodash'
import path from 'path'
import slash from 'slash'

export function isUrl(req: string) {
    return (
        req.startsWith('http://') ||
        req.startsWith('https://') ||
        req.startsWith('//')
    )
}
export interface OptimizeAnalysisResult {
    isCommonjs: { [name: string]: true }
}

export class Lock extends EventEmitter {
    READY_EVENT = 'READY_EVENT'
    isReady = false
    constructor() {
        super()
        this.once(this.READY_EVENT, () => {
            this.isReady = true
        })
    }
    ready() {
        this.emit(this.READY_EVENT)
    }
    wait(): Promise<any> {
        return once(this, this.READY_EVENT)
    }
}

export function osAgnosticPath(absPath: string | undefined, root: string) {
    if (!root) {
        throw new Error(
            `root argument is required, cannot make os agnostic path for ${absPath}`,
        )
    }
    if (!absPath) {
        return ''
    }
    if (!path.isAbsolute(absPath)) {
        absPath = path.resolve(root, absPath)
    }
    return slash(path.relative(root, absPath))
}

export function unique<T>(array: T[], key = (x: T): any => x): T[] {
    const cache: Record<any, boolean> = {}
    return array.filter(function(a) {
        const keyed = key(a)
        if (!cache[keyed]) {
            cache[keyed] = true
            return true
        }
        return false
    }, {})
}

// namespace:/path/to/file -> /path/to/file
export function stripColon(input?: string) {
    if (!input) {
        return ''
    }
    const index = input.indexOf(':')
    if (index === -1) {
        return input
    }
    const clean = input.slice(index + 1)
    return clean
}

function convertKeys<T>(obj: T, cb: (k: string) => string): T {
    const x: T = Array.isArray(obj) ? ([] as any) : {}

    forOwn(obj, (v, k) => {
        if (isPlainObject(v) || Array.isArray(v)) v = convertKeys(v, cb)

        x[cb(k)] = v
    })

    return x
}

export function runFunctionOnPaths(
    x: Metadata,
    func: (x: string) => string = stripColon,
): Metadata {
    x = convertKeys(x, func)
    for (const input in x.inputs) {
        const v = x.inputs[input]
        x.inputs[input] = {
            ...v,
            imports: v.imports
                ? v.imports.map((x) => ({ path: func(x.path) }))
                : [],
        }
    }
    for (const output in x.outputs) {
        const v = x.outputs[output]
        x.outputs[output] = {
            ...v,
            imports: v.imports
                ? v.imports.map((x) => ({ path: func(x.path) }))
                : [],
        }
    }

    return x
}
