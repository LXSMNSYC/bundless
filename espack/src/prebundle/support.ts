import { Metadata } from 'esbuild'
import { EventEmitter, once } from 'events'
import { forOwn, isPlainObject } from 'lodash'
import path from 'path'
import slash from 'slash'

export function isUrl(req: string) {
    return req.startsWith('http://') || req.startsWith('https://')
}

// TODO import OptimizeAnalysisResult from vite
export interface OptimizeAnalysisResult {
    isCommonjs: { [name: string]: true }
}

export class Lock extends EventEmitter {
    READY_EVENT = 'READY_EVENT'
    isReady = false
    constructor() {
        super()
        this.once(this.READY_EVENT, () => {
            this.isReady = true
        })
    }
    ready() {
        this.emit(this.READY_EVENT)
    }
    wait(): Promise<any> {
        return once(this, this.READY_EVENT)
    }
}

export function osAgnosticPath(absPath?: string | undefined) {
    if (!absPath) {
        return ''
    }
    if (!path.isAbsolute(absPath)) {
        absPath = path.resolve(absPath)
    }
    return slash(path.relative(process.cwd(), absPath))
}

export function unique<T>(array: T[], key = (x: T): any => x): T[] {
    const cache: Record<any, boolean> = {}
    return array.filter(function(a) {
        const keyed = key(a)
        if (!cache[keyed]) {
            cache[keyed] = true
            return true
        }
        return false
    }, {})
}

export function stripColon(input?: string) {
    if (!input) {
        return ''
    }
    const index = input.indexOf(':')
    if (index === -1) {
        return input
    }
    const clean = input.slice(index + 1)
    return clean
}

function convertKeys<T>(obj: T, cb: (k: string) => string): T {
    const x: T = Array.isArray(obj) ? ([] as any) : {}

    forOwn(obj, (v, k) => {
        if (isPlainObject(v) || Array.isArray(v)) v = convertKeys(v, cb)

        x[cb(k)] = v
    })

    return x
}

export function removeColonsFromMeta(x: Metadata): Metadata {
    x = convertKeys(x, stripColon)
    for (const input in x.inputs) {
        const v = x.inputs[input]
        x.inputs[input] = {
            ...v,
            imports: v.imports.map((x) => ({ path: stripColon(x.path) })),
        }
    }
    for (const output in x.outputs) {
        const v = x.outputs[output]
        x.outputs[output] = {
            ...v,
            imports: v.imports.map((x) => ({ path: stripColon(x.path) })),
        }
    }

    return x
}
