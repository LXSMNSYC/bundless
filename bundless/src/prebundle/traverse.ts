import {
    HtmlIngestPlugin,
    NodeModulesPolyfillPlugin,
    NodeResolvePlugin,
} from '../plugins'
import deepmerge from 'deepmerge'
import { build, BuildOptions, Metadata, Plugin } from 'esbuild'
import { promises as fsp } from 'fs'
import fsx, { copySync } from 'fs-extra'
import os from 'os'
import path from 'path'
import slash from 'slash'
import {
    importableAssets,
    isRunningWithYarnPnp,
    JS_EXTENSIONS,
    MAIN_FIELDS,
} from '../constants'

import { osAgnosticPath, runFunctionOnPaths } from './support'
import fromEntries from 'fromentries'
import { stripColon, unique } from './support'
import { flatten } from '../utils'
import { logger } from '../logger'
import {
    commonEsbuildOptions,
    generateDefineObject,
    resolvableExtensions,
} from './esbuild'

type Args = {
    esbuildCwd: string
    root: string
    entryPoints: string[]
    plugins: Plugin[]
    esbuildOptions?: Partial<BuildOptions>
    // resolver?: (cwd: string, id: string) => string
    stopTraversing?: (resolvedPath: string) => boolean
}

export async function traverseWithEsbuild({
    entryPoints,
    esbuildCwd,
    root,
    plugins,
    esbuildOptions = { plugins: [] },
    stopTraversing,
}: Args): Promise<string[]> {
    const destLoc = await fsp.realpath(
        path.resolve(await fsp.mkdtemp(path.join(os.tmpdir(), 'dest'))),
    )

    for (let entry of entryPoints) {
        if (!path.isAbsolute(entry)) {
            throw new Error(
                `All entryPoints of traverseWithEsbuild must be absolute: ${entry}`,
            )
        }
    }

    try {
        const metafile = path.join(destLoc, 'meta.json')
        // logger.log(`Running esbuild in cwd '${process.cwd()}'`)

        await build(
            deepmerge(
                {
                    ...commonEsbuildOptions,
                    define: generateDefineObject({}),
                    entryPoints,
                    outdir: destLoc,
                    metafile,
                    plugins: [
                        ...(plugins || []),
                        HtmlIngestPlugin({ root }),
                        ExternalButInMetafile(),
                        NodeModulesPolyfillPlugin(),
                        NodeResolvePlugin({
                            name: 'traverse-node-resolve',
                            mainFields: MAIN_FIELDS,
                            extensions: resolvableExtensions,
                            onResolved: function external(resolved) {
                                if (
                                    stopTraversing &&
                                    stopTraversing(resolved)
                                ) {
                                    return {
                                        namespace: externalNamespace,
                                        path: resolved,
                                    }
                                }
                                return
                            },
                            onNonResolved: (p) => {
                                console.error(
                                    `Cannot resolve '${p}' during traversal`,
                                )
                                // return {
                                //     external: true,
                                // }
                            },
                        }),
                    ],
                } as BuildOptions,
                esbuildOptions,
            ),
        )
        let meta: Metadata = JSON.parse(
            await (await fsp.readFile(metafile)).toString(),
        )
        // console.log(JSON.stringify(meta, null, 4))
        meta = runFunctionOnPaths(meta, stripColon)

        const res = metaToTraversalResult({
            meta,
            entryPoints,
            root,
            esbuildCwd,
        })

        return Object.keys(res)
    } finally {
        await fsx.remove(destLoc)
    }
}

const externalNamespace = 'external-but-keep-in-metafile'
function ExternalButInMetafile(): Plugin {
    return {
        name: externalNamespace,
        setup(build) {
            const externalModule = 'externalModuleXXX'
            build.onResolve(
                {
                    filter: new RegExp(externalModule),
                    namespace: externalNamespace,
                },
                (args) => {
                    if (args.path !== externalModule) {
                        return
                    }
                    return {
                        external: true,
                    }
                },
            )
            build.onLoad(
                {
                    filter: /.*/,
                    namespace: externalNamespace,
                },
                (args) => {
                    const contents = `export * from '${externalModule}'`
                    return {
                        contents,
                        loader: 'js',
                        // resolveDir: path.dirname(args.path),
                    }
                },
            )
        },
    }
}

type TraversalGraph = Record<string, string[]>

/**
 * Returns a module graph implemented as an object, keys are modules (relative paths from root), values are arrays of key's imports (absolute paths)
 */
export function metaToTraversalResult({
    meta,
    entryPoints,
    esbuildCwd,
    root,
}: {
    meta: Metadata
    esbuildCwd: string
    root: string
    entryPoints: string[]
}): TraversalGraph {
    if (!path.isAbsolute(esbuildCwd)) {
        throw new Error('esbuildCwd must be an absolute path')
    }
    for (let entry of entryPoints) {
        if (!path.isAbsolute(entry)) {
            throw new Error('entry must be an absolute path')
        }
    }
    const alreadyProcessed = new Set<string>()
    // must be all absolute paths
    let toProcess = entryPoints
    const result: TraversalGraph = {}
    // abs path -> input info
    const inputs: Record<string, { imports: { path: string }[] }> = fromEntries(
        Object.keys(meta.inputs).map((k) => {
            const abs = path.resolve(esbuildCwd, k)
            return [abs, meta.inputs[k]]
        }),
    )
    while (toProcess.length) {
        const newImports = flatten(
            toProcess.map((absPath): string[] => {
                if (alreadyProcessed.has(absPath)) {
                    return []
                }
                alreadyProcessed.add(absPath)
                // newEntry = path.posix.normalize(newEntry) // TODO does esbuild always use posix?
                const input = inputs[absPath]
                if (input == null) {
                    throw new Error(
                        `entry '${absPath}' is not present in esbuild metafile inputs ${JSON.stringify(
                            Object.keys(inputs),
                            null,
                            2,
                        )}`,
                    )
                }
                // abs paths
                const currentImports: string[] = input.imports
                    ? input.imports
                          .map((x) => x.path)
                          .map((x) => {
                              if (!path.isAbsolute(x)) {
                                  return path.resolve(esbuildCwd, x)
                              }
                              return x
                          })
                          .filter((x) => Boolean(x))
                    : []
                // newImports.push(...currentImports)

                const importer = osAgnosticPath(
                    path.resolve(esbuildCwd, absPath),
                    root,
                )
                if (!result[importer]) {
                    result[importer] = []
                }
                for (let importee of currentImports) {
                    if (!importee) {
                        continue
                    }
                    importee = osAgnosticPath(importee, root)
                    result[importer].push(importee)
                }

                return currentImports
            }),
        ).filter(Boolean)
        toProcess = newImports
    }
    return result
    // find the right output getting the key of the right output.inputs == input
    // get the imports of the inputs.[entry].imports and attach them the importer
    // do the same with the imports just found
    // return the list of input files
}
