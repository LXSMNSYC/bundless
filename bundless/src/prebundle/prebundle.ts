import fs from 'fs-extra'
import path from 'path'
import slash from 'slash'
import { COMMONJS_ANALYSIS_PATH } from '../constants'
import { logger } from '../logger'
import { needsPrebundle } from '../utils'
import { bundleWithEsBuild } from './esbuild'
import { printStats } from './stats'
import { osAgnosticPath } from './support'
import { traverseWithEsbuild } from './traverse'

export async function prebundle({ entryPoints, filter, root, dest }) {
    const dependenciesPaths = await getDependenciesPaths({ entryPoints, root, filter })
    logger.log(
        `prebundling [${dependenciesPaths
            .map((x) => osAgnosticPath(x, root))
            .join(', ')}]`,
    )
    await fs.remove(dest)
    const { bundleMap, analysis, stats } = await bundleWithEsBuild({
        dest,
        root,
        entryPoints: dependenciesPaths.map((x) => path.resolve(root, x)),
    })

    const analysisFile = path.join(dest, COMMONJS_ANALYSIS_PATH)
    await fs.createFile(analysisFile)
    // console.log({ analysis })
    await fs.writeFile(analysisFile, JSON.stringify(analysis, null, 4))

    console.info(printStats(stats))
    return bundleMap
}

export async function getDependenciesPaths({ entryPoints, filter, root }) {
    // serve react refresh runtime
    const traversalResult = await traverseWithEsbuild({
        entryPoints,
        root,
        stopTraversing: filter,
        esbuildCwd: process.cwd(),
    })
    let resolvedFiles = traversalResult
        .map((x) => {
            return x.resolvedImportPath
        })
        .filter(Boolean)
        .filter((x) => filter(x))
        .map((x) => slash(path.relative(root, x)))
    resolvedFiles = Array.from(new Set(resolvedFiles))
    return resolvedFiles
}

// on start, check if already optimized dependencies, else
// traverse from entrypoints and get all imported paths, stopping when finding a node_module
// start bundling these modules and store them in a web_modules folder
// save a commonjs modules list in web_modules folder
// this can be a plugin that start building when finding a new node_modules plugin that should not be there
// if it sees a path that has a node_modules inside it blocks the server, start bundling and restart everything
