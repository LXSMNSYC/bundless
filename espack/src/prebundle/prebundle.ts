import chalk from 'chalk'
import fs, { stat } from 'fs-extra'
import path from 'path'
import slash from 'slash'
import { RawSourceMap } from 'source-map'
import { COMMONJS_ANALYSIS_PATH } from '../constants'
import { PluginHooks } from '../plugin'
import { isNodeModule, readFile } from '../utils'
import { bundleWithEsBuild } from './esbuild'
import { printStats } from './stats'
import { traverseWithEsbuild } from './traverse'

const debug = require('debug')('esbuild')

export async function prebundle({ entryPoints, root, dest }) {
    const dependenciesPaths = await getDependenciesPaths({ entryPoints, root })

    await fs.remove(dest)
    const { bundleMap, analysis, stats } = await bundleWithEsBuild({
        dest,
        root,
        entryPoints: dependenciesPaths.map((x) => path.resolve(root, x)),
    })

    const analysisFile = path.join(dest, COMMONJS_ANALYSIS_PATH)
    await fs.createFile(analysisFile)
    // console.log({ analysis })
    await fs.writeFile(analysisFile, JSON.stringify(analysis, null, 4))

    console.info(printStats(stats))
    return bundleMap
}

export async function getDependenciesPaths({ entryPoints, root }) {
    // serve react refresh runtime
    const traversalResult = await traverseWithEsbuild({
        entryPoints,
        stopTraversing: isNodeModule,
        cwd: root,
    })
    let resolvedFiles = traversalResult
        .map((x) => {
            return x.resolvedImportPath
        })
        .filter(Boolean)
        .filter((x) => isNodeModule(x))
        .map((x) => slash(path.relative(root, x)))
    resolvedFiles = Array.from(new Set(resolvedFiles))
    return resolvedFiles
}

// on start, check if already optimized dependencies, else
// traverse from entrypoints and get all imported paths, stopping when finding a node_module
// start bundling these modules and store them in a web_modules folder
// save a commonjs modules list in web_modules folder
// this can be a plugin that start building when finding a new node_modules plugin that should not be there
// if it sees a path that has a node_modules inside it blocks the server, start bundling and restart everything
