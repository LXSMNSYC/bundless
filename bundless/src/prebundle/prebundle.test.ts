import memoize from 'micro-memoize'
import path from 'path'
import { getDependenciesPaths } from './prebundle'
import { traverseWithEsbuild } from './traverse'

test('traverseWithEsbuild', async () => {
    const entry = path.resolve('fixtures/with-many-dependencies/index.html')
    const deps = await traverseWithEsbuild({
        entryPoints: [entry],
        esbuildCwd: process.cwd(),
        stopTraversing: (x) => x.includes('node_module'),
        root: path.dirname(entry),
    })
    expect(deps).toMatchSnapshot()
})

test('memoize', () => {
    let i = 0
    const fn = memoize((x) => {
        return i++
    })
    fn(1)
    fn(1)
    fn.cache.keys = []
    fn.cache.values = []
    fn(1)
    fn(1)
    fn(1)
    expect(i).toBe(2)
})
