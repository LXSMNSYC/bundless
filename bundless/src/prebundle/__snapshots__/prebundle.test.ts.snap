// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`traverseWithEsbuild 1`] = `
Array [
  "index.html",
  "main.js",
  "../../node_modules/slash/index.js",
  "node_modules/react/index.js",
  "node_modules/preact/hooks/dist/hooks.module.js",
  "../../node_modules/react-dom/index.js",
]
`;
